File: server-python/src/revturbine/core/state/impression_history_stores.py
"""Impression history store implementations — Python port of
@revt-eng/core/state/impression-history-stores.ts.

- ``InMemoryImpressionStore`` — ephemeral, in-process.
- ``StorageImpressionStore`` — backed by ``RevTurbineStorage``
  (suitable for any sync storage backend).

Source: revturbine-scaffold/src/core/state/impression-history-stores.ts
"""

from __future__ import annotations

import contextlib
import json
import time
from typing import Any

from revturbine.core.state.impression_history_types import (
    TERMINAL_OUTCOMES,
    ImpressionHistoryStore,
    ImpressionQuery,
    ImpressionRecord,
)
from revturbine.core.state.storage import RevTurbineStorage

__all__ = ["InMemoryImpressionStore", "StorageImpressionStore"]


# ── Shared helpers ──────────────────────────────────────────────────────────


def _apply_query(
    records: list[ImpressionRecord],
    query: ImpressionQuery,
) -> list[ImpressionRecord]:
    """Filter ``records`` by placement IDs, outcomes, and ``since`` timestamp.

    Source: impression-history-stores.ts:147-165
    """
    filtered = records

    placement_ids = query.get("placement_ids")
    if placement_ids:
        ids_set = set(placement_ids)
        filtered = [r for r in filtered if r["placement_id"] in ids_set]

    outcomes = query.get("outcomes")
    if outcomes:
        outcomes_set = set(outcomes)
        filtered = [r for r in filtered if r["outcome"] in outcomes_set]

    since = query.get("since")
    if since:
        since_ms = _parse_iso_to_ms(since)
        filtered = [r for r in filtered if _parse_iso_to_ms(r["occurred_at"]) >= since_ms]

    return filtered


def _extract_retired_ids(records: list[ImpressionRecord]) -> set[str]:
    """Source: impression-history-stores.ts:168-176"""
    return {r["placement_id"] for r in records if r["outcome"] in TERMINAL_OUTCOMES}


def _extract_suppressed_placements(records: list[ImpressionRecord]) -> dict[str, str]:
    """Walk newest → oldest so the first hit per placement wins; drop any
    suppression whose ``suppressUntil`` is in the past.

    Source: impression-history-stores.ts:183-198
    """
    suppressed: dict[str, str] = {}
    now = int(time.time() * 1000)
    for record in reversed(records):
        if record["outcome"] != "suppressed":
            continue
        if record["placement_id"] in suppressed:
            continue
        metadata = record.get("metadata") or {}
        until = metadata.get("suppressUntil")
        if not isinstance(until, str):
            continue
        if _parse_iso_to_ms(until) > now:
            suppressed[record["placement_id"]] = until
    return suppressed


def _parse_iso_to_ms(iso: str) -> int:
    """Mirror JS's ``new Date(iso).getTime()``: returns ms since epoch.

    Tolerates the trailing ``Z`` form; falls back to ``0`` for unparseable
    input (matches JS's NaN → comparator-falsy semantics where the record
    is effectively excluded).
    """
    from datetime import datetime

    try:
        # Python's fromisoformat doesn't accept 'Z' until 3.11; normalize.
        normalized = iso.replace("Z", "+00:00") if iso.endswith("Z") else iso
        return int(datetime.fromisoformat(normalized).timestamp() * 1000)
    except (TypeError, ValueError):
        return 0


# ── In-memory store ─────────────────────────────────────────────────────────


class InMemoryImpressionStore(ImpressionHistoryStore):
    """Ephemeral, in-process impression history store. Suitable for SSR /
    testing.

    Source: impression-history-stores.ts:26-54
    """

    def __init__(self) -> None:
        self._records: dict[str, list[ImpressionRecord]] = {}

    def append(self, user_id: str, record: ImpressionRecord) -> None:
        self._records.setdefault(user_id, []).append(record)

    def query(
        self,
        user_id: str,
        query: ImpressionQuery | None = None,
    ) -> list[ImpressionRecord]:
        records = self._records.get(user_id, [])
        if query:
            records = _apply_query(records, query)
        # Reverse so most-recent is first, matching the TS slice().reverse().
        return list(reversed(records))

    def get_retired_placement_ids(self, user_id: str) -> set[str]:
        return _extract_retired_ids(self._records.get(user_id, []))

    def get_suppressed_placements(self, user_id: str) -> dict[str, str]:
        return _extract_suppressed_placements(self._records.get(user_id, []))

    def clear(self, user_id: str) -> None:
        self._records.pop(user_id, None)


# ── Storage-backed store ────────────────────────────────────────────────────


_STORAGE_PREFIX = "revturbine:impression-history"
_DEFAULT_MAX_RECORDS = 500


class StorageImpressionStore(ImpressionHistoryStore):
    """Impression history store backed by any ``RevTurbineStorage``.

    Records are serialized as a JSON array per user key. A configurable
    ``max_records`` limit prevents unbounded growth (default: 500).

    Source: impression-history-stores.ts:73-141
    """

    def __init__(
        self,
        *,
        storage: RevTurbineStorage,
        tenant_id: str,
        max_records: int = _DEFAULT_MAX_RECORDS,
    ) -> None:
        self._storage = storage
        self._tenant_id = tenant_id
        self._max_records = max_records

    def append(self, user_id: str, record: ImpressionRecord) -> None:
        records = self._load(user_id)
        records.append(record)
        if len(records) > self._max_records:
            del records[0 : len(records) - self._max_records]
        self._save(user_id, records)

    def query(
        self,
        user_id: str,
        query: ImpressionQuery | None = None,
    ) -> list[ImpressionRecord]:
        records = self._load(user_id)
        if query:
            records = _apply_query(records, query)
        return list(reversed(records))

    def get_retired_placement_ids(self, user_id: str) -> set[str]:
        return _extract_retired_ids(self._load(user_id))

    def get_suppressed_placements(self, user_id: str) -> dict[str, str]:
        return _extract_suppressed_placements(self._load(user_id))

    def clear(self, user_id: str) -> None:
        self._storage.remove_item(self._key(user_id))

    def _key(self, user_id: str) -> str:
        return f"{_STORAGE_PREFIX}:{self._tenant_id}:{user_id}"

    def _load(self, user_id: str) -> list[ImpressionRecord]:
        raw = self._storage.get_item(self._key(user_id))
        if not raw:
            return []
        try:
            parsed: Any = json.loads(raw)
        except (json.JSONDecodeError, ValueError):
            return []
        return parsed if isinstance(parsed, list) else []

    def _save(self, user_id: str, records: list[ImpressionRecord]) -> None:
        with contextlib.suppress(Exception):
            self._storage.set_item(self._key(user_id), json.dumps(records))

File: server-rust/src/state/impression_history_stores.rs
//! Impression-history store implementations.
//!
//! [`InMemoryImpressionStore`] is ephemeral and in-process — the backend the
//! headless SDK ships. A durable store is a caller decision.
//!
//! Source: revturbine-scaffold/src/core/state/impression-history-stores.ts

use std::collections::{HashMap, HashSet};

use crate::state::impression_history_types::{
    parse_iso_to_ms, ImpressionHistoryStore, ImpressionQuery, ImpressionRecord,
};

/// Filter records by placement ids, outcomes, and a `since` floor.
///
/// Source: impression-history-stores.ts:147-165 (applyQuery)
#[must_use]
pub fn apply_query(records: &[ImpressionRecord], query: &ImpressionQuery) -> Vec<ImpressionRecord> {
    let mut filtered: Vec<ImpressionRecord> = records.to_vec();

    if let Some(ids) = query.placement_ids.as_ref().filter(|v| !v.is_empty()) {
        let set: HashSet<&str> = ids.iter().map(String::as_str).collect();
        filtered.retain(|r| set.contains(r.placement_id.as_str()));
    }
    if let Some(outcomes) = query.outcomes.as_ref().filter(|v| !v.is_empty()) {
        let set: HashSet<&str> = outcomes.iter().map(String::as_str).collect();
        filtered.retain(|r| set.contains(r.outcome.as_str()));
    }
    if let Some(since) = query.since.as_ref().filter(|s| !s.is_empty()) {
        let since_ms = parse_iso_to_ms(since);
        filtered.retain(|r| parse_iso_to_ms(&r.occurred_at) >= since_ms);
    }

    filtered
}

/// Placements permanently retired — those with a terminal outcome.
///
/// Source: impression-history-stores.ts:168-176 (extractRetiredIds)
#[must_use]
pub fn extract_retired_ids(records: &[ImpressionRecord]) -> HashSet<String> {
    records
        .iter()
        .filter(|r| r.is_terminal())
        .map(|r| r.placement_id.clone())
        .collect()
}

/// Placements currently inside a suppression window, mapped to its expiry.
///
/// Walks **newest → oldest** so the most recent window per placement wins, and
/// drops any whose `suppressUntil` has already passed. A cooldown is *any*
/// record carrying a future `suppressUntil` — which since plan 167 Q-1
/// includes `dismissed` and bare `clicked_thru`, not only `suppressed`.
///
/// Source: impression-history-stores.ts:183-198 (extractSuppressedPlacements)
#[must_use]
pub fn extract_suppressed_placements(
    records: &[ImpressionRecord],
    now_ms: i64,
) -> HashMap<String, String> {
    let mut suppressed: HashMap<String, String> = HashMap::new();
    for record in records.iter().rev() {
        if suppressed.contains_key(&record.placement_id) {
            continue;
        }
        let Some(until) = record
            .metadata
            .as_ref()
            .and_then(|m| m.get("suppressUntil"))
            .and_then(|v| v.as_str())
        else {
            continue;
        };
        if parse_iso_to_ms(until) > now_ms {
            suppressed.insert(record.placement_id.clone(), until.to_string());
        }
    }
    suppressed
}

fn system_now_ms() -> i64 {
    use std::time::{SystemTime, UNIX_EPOCH};
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0)
}

/// Ephemeral, in-process impression history.
///
/// Carries its own clock so suppression-window expiry is testable without
/// sleeping — the trait signature stays identical to the TS and Python stores.
///
/// Source: impression-history-stores.ts:26-54
pub struct InMemoryImpressionStore {
    records: HashMap<String, Vec<ImpressionRecord>>,
    now_fn: Box<dyn Fn() -> i64 + Send + Sync>,
}

impl Default for InMemoryImpressionStore {
    fn default() -> Self {
        Self::new()
    }
}

impl InMemoryImpressionStore {
    /// An empty store on the system clock.
    #[must_use]
    pub fn new() -> Self {
        Self {
            records: HashMap::new(),
            now_fn: Box::new(system_now_ms),
        }
    }

    /// An empty store on an explicit clock.
    #[must_use]
    pub fn with_clock(now_fn: Box<dyn Fn() -> i64 + Send + Sync>) -> Self {
        Self {
            records: HashMap::new(),
            now_fn,
        }
    }
}

impl ImpressionHistoryStore for InMemoryImpressionStore {
    fn append(&mut self, user_id: &str, record: ImpressionRecord) {
        self.records
            .entry(user_id.to_string())
            .or_default()
            .push(record);
    }

    fn query(&self, user_id: &str, query: Option<&ImpressionQuery>) -> Vec<ImpressionRecord> {
        let records = self.records.get(user_id).cloned().unwrap_or_default();
        let mut filtered = match query {
            Some(q) => apply_query(&records, q),
            None => records,
        };
        // Most-recent first, matching the TS `slice().reverse()`.
        filtered.reverse();
        filtered
    }

    fn get_retired_placement_ids(&self, user_id: &str) -> HashSet<String> {
        self.records
            .get(user_id)
            .map(|r| extract_retired_ids(r))
            .unwrap_or_default()
    }

    fn get_suppressed_placements(&self, user_id: &str) -> HashMap<String, String> {
        self.records
            .get(user_id)
            .map(|r| extract_suppressed_placements(r, (self.now_fn)()))
            .unwrap_or_default()
    }

    fn clear(&mut self, user_id: &str) {
        self.records.remove(user_id);
    }
}
